dlt/transformations/transformation.py

```diff
@@ -0,0 +1,194 @@
+    write_disposition: TWriteDisposition = None,
+    columns: TTableSchemaColumns = None,
+    primary_key: TColumnNames = None,
+    merge_key: TColumnNames = None,
+    schema_contract: TSchemaContract = None,
+    table_format: TTableFormat = None,
+    references: TTableReferenceParam = None,
+    selected: bool = True,
+    spec: Type[TransformConfiguration] = None,
+    parallelized: bool = False,
+    incremental: Optional[TIncrementalConfig] = None,
+) -> DltTransformResource:
+    # resolve defaults etc
+
+    resource_name = name if name and not callable(name) else get_callable_name(func)
+
+    # check function type, for generators we assume a regular resource
+    is_regular_resource = inspect.isgeneratorfunction(
+        inspect.unwrap(func)
+    ) or inspect.isgeneratorfunction(func)
+
+    # build transformation function
+    def transformation_function(*args: Any, **kwargs: Any) -> Iterator[TDataItems]:
+        # NOTE: this is most likely not correct
+        config = resolve_configuration(
+            TransformConfiguration(),
+            sections=(known_sections.EXTRACT, resource_name),
+            accept_partial=False,
+        )
+
+        # collect all datasets from args
+        datasets: List[ReadableDBAPIDataset] = []
+        for arg in args:
+            if isinstance(arg, ReadableDBAPIDataset):
+                datasets.append(arg)
+
+        # NOTE: there may be cases where some other dataset is used to get a starting
+        # point and it will be on a different destination.
+        if datasets:
+            for d in datasets:
+                if not d.is_same_physical_destination(datasets[0]):
+                    raise IncompatibleDatasetsException(
+                        resource_name,
+                        "All datasets used in transformation must be on the"
+                        + " same physical destination.",
+                    )
+        else:
+            raise IncompatibleDatasetsException(
+                resource_name,
+                "No datasets detected in transformation. Please supply all used datasets via"
+                " transform function arguments.",
+            )
+
+        # Determine wether we use sql (model) or python (arrow_iterator) transformation
+        # we need to supply the curent schema name (=source name)to the dataset constructor
+
+        resolved_transformation_type = "python"
+        try:
+            schema_name = dlt.current.source().name
+            current_pipeline = dlt.current.pipeline()
+            if not current_pipeline._destination:
+                pass
+            elif datasets[0].is_same_physical_destination(
+                dlt.current.pipeline().dataset(schema=schema_name)
+            ):
+                resolved_transformation_type = "model"
+        # if we cannot reach the destination, or a running outside of a pipeline, we extract frames
+        except (PipelineConfigMissing, CurrentSourceNotAvailable):
```

Comment:
if we cannot find source context or pipeline context we should IMO return model, not frames. so you can still do `pipeline.extract()` when pipeline does not define destination at all. (this is a valid use case)

Reply:
Hm, ok, I am not sure, but I have changed it like this now.